contracts/DepositVault.sol

```diff
@@ -25,53 +27,73 @@ contract DepositVault is EIP712 {
 
-    function deposit() public payable {
-        require(msg.value > 0, "Deposit amount must be greater than 0");
-
-        uint256 depositIndex = deposits.length;
-        deposits.push(Deposit(payable(msg.sender), msg.value));
-        emit DepositMade(msg.sender, depositIndex, msg.value);
+    function deposit(uint256 amount, address tokenAddress) public payable {
+        require(amount > 0 || msg.value > 0, "Deposit amount must be greater than 0");
+        if(msg.value > 0){
+            require(tokenAddress == address(0), "Token address must be 0x0 for ETH deposits");
+            uint256 depositIndex = deposits.length;
+            deposits.push(Deposit(payable(msg.sender), msg.value, tokenAddress));
+            emit DepositMade(msg.sender, depositIndex, msg.value, tokenAddress);
+        } else {
+            require(tokenAddress != address(0), "Token address must not be 0x0 for token deposits");
+            IERC20 token = IERC20(tokenAddress);
+            token.transferFrom(msg.sender, address(this), amount);
+            uint256 depositIndex = deposits.length;
+            deposits.push(Deposit(payable(msg.sender), amount, tokenAddress));
+            emit DepositMade(msg.sender, depositIndex, amount, tokenAddress);
+
+        }
     }
 
     function getWithdrawalHash(Withdrawal memory withdrawal) public view returns (bytes32) {
         return _hashTypedDataV4(keccak256(abi.encode(WITHDRAWAL_TYPEHASH, withdrawal.amount, withdrawal.nonce)));
     }
 
-    function withdraw(uint256 amount, uint256 nonce, bytes memory signature, address payable recipient) public {
+    function withdraw(uint256 amount, uint256 nonce, bytes memory signature, address payable recipient, address tokenAddress) public {
         require(nonce < deposits.length, "Invalid deposit index");
         Deposit storage depositToWithdraw = deposits[nonce];
         bytes32 withdrawalHash = getWithdrawalHash(Withdrawal(amount, nonce));
         address signer = withdrawalHash.recover(signature);
         require(signer == depositToWithdraw.depositor, "Invalid signature");
         require(!usedWithdrawalHashes[withdrawalHash], "Withdrawal has already been executed");
         require(amount == depositToWithdraw.amount, "Withdrawal amount must match deposit amount");
+        require(tokenAddress == depositToWithdraw.tokenAddress, "Withdrawal token address must match deposit token address");
 
         usedWithdrawalHashes[withdrawalHash] = true;
         depositToWithdraw.amount = 0;
-        recipient.transfer(amount);
+        
+        if(tokenAddress == address(0)){
+            recipient.transfer(amount);
+        } else {
+            IERC20 token = IERC20(tokenAddress);
+            token.transfer(recipient, amount);
+        }
 
         emit WithdrawalMade(recipient, amount);
     }
 
-    function withdrawDeposit(uint256 depositIndex) public {
+    function withdrawDeposit(uint256 depositIndex, address tokenAddress) public {
```

Comment:
same here for the token address, adding this extra parameter would make it more cumbersome, we just need the index value